### PR TITLE
Perf of dereference is unstable

### PR DESCRIPTION
**Pull request by P0lip**
See https://github.com/APIDevTools/json-schema-ref-parser/issues/211

### CODE REVIEW

File: lib/dereference.js [hunk 1]
```diff
@@ -28,65 +28,61 @@ function dereference (parser, options) {
...
  */
-function crawl (obj, path, pathFromRoot, parents, processedObjects, dereferencedCache, $refs, options) {
+function crawl (obj, path, pathFromRoot, parents, dereferencedCache, $refs, options) {
   let dereferenced;
   let result = {
     value: obj,
     circular: false
   };
 
-  if (options.dereference.circular === "ignore" || processedObjects.indexOf(obj) === -1) {
```
Comment (wmhilton) [inline lib/dereference.js:44]: Wait, so you fixed the performance by removing protection against circular references?

Comment (P0lip) [reply lib/dereference.js:44]: Nope! 
It's just a redundant check that didn't seem to have any actual impact on perf.

https://github.com/stoplightio/json-schema-ref-parser/blob/55dee7ed978b734a1fd354e01197236b3eaa106c/lib/dereference.js#L108L126 is what actually made impact. That added check didn't contribute.

For a context this is the change I'm talking about https://github.com/stoplightio/json-schema-ref-parser/commit/5c3eece30a212cda80212244d904bd63792e545c
I just didn't notice `options.dereference.circular === "ignore" || processedObjects.indexOf(obj) === -1)` changing anything so I reverted to the initial version of this code (prior to the aforementioned commit)

Comment (P0lip) [reply lib/dereference.js:44]: We may potentially just use a bit more conservative change that was implemented here https://github.com/APIDevTools/json-schema-ref-parser/pull/212
It only swapped arrays with sets, but this already mitigated the perf cliff.

**Comment by P0lip** [comment]
https://github.com/stoplightio/platform-internal/issues/5517 reminded me about this 
Since the URL has `mid`, and masking = dereferenced output, they might be encountering this issue.
This would actually align with what Em said in the linked ticket - the whole process took a bit over a minute in case of the file he dealt with.

**Comment by chohmann** [comment]
@P0lip if https://github.com/stoplightio/platform-internal/issues/5517 is caused by this, can I assign that ticket to you guys?

**Review by domagojk — APPROVED** [review]
It looks good. If it was a redundant check then can't complain about removing it :)
Just remove a jsdoc comment for `processedObjects`, but it's approve from me 👍

File: lib/dereference.js [hunk 2]
```diff
@@ -104,7 +100,7 @@ function crawl (obj, path, pathFromRoot, parents, processedObjects, dereferenced
  * @param {$RefParserOptions} options
  * @returns {{value: object, circular: boolean}}
  */
-function dereference$Ref ($ref, path, pathFromRoot, parents, processedObjects, dereferencedCache, $refs, options) {
+function dereference$Ref ($ref, path, pathFromRoot, parents, dereferencedCache, $refs, options) {
```
[the remark this answers is not in the document]
Comment (P0lip) [reply lib/dereference.js:103]: Yeah, in fact I removed an incorrect comment.

**Pull request merged**